File: brickwall.py
from sys import stdin,stdout,setrecursionlimit

def solve(c1,c2,c3,curr):
    global ans,gaps,H,visited
    visited[c1][c2][c3] = True
    if ans == "YES" or curr > H:
        return
    if curr == H:
        ans = "YES"
        return
    if not gaps[curr + 1] and c1 > 0 and not visited[c1 - 1][c2][c3]:
        solve(c1 - 1,c2,c3,curr + 1)
    if not gaps[curr + 2] and c2 > 0 and not visited[c1][c2 - 1][c3]:
        solve(c1,c2 - 1,c3,curr + 2)
    if not gaps[curr + 3] and c3 > 0 and not visited[c1][c2][c3 - 1]:
        solve(c1,c2,c3 - 1,curr + 3)

#maximum length is 300
gaps,ans = [False for i in range(305)],"NO"
N,c1,c2,c3 = map(int,stdin.readline().split())
row = list(map(int,stdin.readline().split()))
#gaps contain the distances of every connection measured from the left
H = 0
for i in range(N - 1):
    H += row[i]
    gaps[H] = True
H += row[N - 1]
#use visited 3D array to prevent revisiting same state again
visited = [[[False for i in range(c3 + 1)] for j in range(c2 + 1)] for k in range(c1 + 1)]
solve(c1,c2,c3,0)
stdout.write(f"{ans}\n")

File: brickwall.cpp
#include <bits/stdc++.h>
#include <string>

using namespace std;

const int MAX_N = 303;
string ans = "NO";
bool gaps[MAX_N], visited[MAX_N][MAX_N][MAX_N];
int counter = 0;

void solve(int c1, int c2, int c3, int curr) {
    visited[c1][c2][c3] = true;
    if (ans == "YES" || curr > counter) return;
    if (curr == counter) {
        ans = "YES";
        return;
    }
    if (!gaps[curr + 1] && c1 > 0 && !visited[c1 - 1][c2][c3]) solve(c1 - 1,c2,c3,curr + 1);
    if (!gaps[curr + 2] && c2 > 0 && !visited[c1][c2 - 1][c3]) solve(c1,c2 - 1,c3,curr + 2);
    if (!gaps[curr + 3] && c3 > 0 && !visited[c1][c2][c3 - 1]) solve(c1,c2,c3 - 1,curr +3);
}

int main() {
    int N,c1,c2,c3; scanf("%d %d %d %d",&N,&c1,&c2,&c3);
    for (int i = 0; i < N; i++) {
        int a; scanf("%d",&a);
        counter += a;
        if (i < N  - 1) gaps[counter] = true;
    }
    solve(c1,c2,c3,0);
    printf("%s\n",ans.c_str());
    return 0;
}